cranelift/codegen/src/isa/aarch64/lower_inst.rs

```diff
@@ -1858,33 +1850,36 @@ pub(crate) fn lower_insn_to_regs<C: LowerCtx<I = Inst>>(
 
             let from_ty = ctx.input_ty(insn, 0);
             let to_ty = ctx.output_ty(insn, 0);
-            let from_bits = ty_bits(from_ty);
-            let to_bits = ty_bits(to_ty);
-
-            if from_ty.is_vector() || from_bits > 64 || to_bits > 64 {
-                return Err(CodegenError::Unsupported(format!(
-                    "{}: Unsupported type: {:?}",
-                    op, from_ty
-                )));
-            }
-
-            assert!(from_bits <= to_bits);
+            let from_bits = ty_bits(from_ty) as u8;
+            let to_bits = ty_bits(to_ty) as u8;
 
-            let rd = get_output_reg(ctx, outputs[0]).only_reg().unwrap();
-            let rn = put_input_in_reg(ctx, inputs[0], NarrowValueMode::None);
+            let src = put_input_in_regs(ctx, inputs[0]);
+            let dst = get_output_reg(ctx, outputs[0]);
 
-            if from_bits == to_bits {
-                ctx.emit(Inst::gen_move(rd, rn, to_ty));
-            } else {
-                let to_bits = if to_bits > 32 { 64 } else { 32 };
-                let from_bits = from_bits as u8;
+            if to_bits > from_bits {
+                // If the dst type is larger than src, we need to sign extend `src`.
+                // In case of a b128 type, we only extend the bottom register, later we
+                // copy the bottom register into the top, since for b128 the top register
+                // is always equal to the bottom one.
                 ctx.emit(Inst::Extend {
-                    rd,
-                    rn,
+                    rd: dst.regs()[0],
+                    rn: src.regs()[0],
                     signed: true,
-                    from_bits,
-                    to_bits,
+                    from_bits: cmp::min(from_bits, 64),
+                    to_bits: cmp::min(to_bits, 64),
                 });
+            } else {
+                // Smaller integers/booleans are stored with high-order bits undefined, so we can
```

Comment:
I am afraid that this is not true for vector types; e.g. consider an operation from `B8X8` to `B16X8`.

Unless both the input and the output have the same lane sizes (the number of lanes is always the same, as specified by the operation descriptions), you will have to write much more complicated logic, and will probably be unable to merge the scalar and vector code paths, as you do now.

Reply:
Right, I didn't really prepare this for those types (non 128 bit vectors). I haven't really seen them used anywhere so far.

Would it be acceptable to assert on 128bit vectors for now?

Reply:
No, just keep the block (lines 1864 - 1869) that you have removed. Of course, you will need to change the condition - I suppose to `from_ty.is_vector() && (from_bits != 128 || from_bits != to_bits)`.

IMHO assertions should be used for verifying invariants guaranteed by the IR descriptions, e.g. checking that the number of input lanes is equal to the number of output lanes. That is the approach I have used in my recent changes.

Reply:
Even after your latest changes, this comment should be changed to `Smaller scalar integers/booleans are...`.